testonly/hammer/record.go

```diff
@@ -28,48 +30,99 @@ const (
 	smrCount = 30
 )
 
-// smrStash provides thread-safe access to an ordered, bounded queue of previously
-// witnessed SMRs
-type smrStash struct {
-	mu sync.RWMutex
+// gossipHub allows details of what has been written to and read from the Map
+// to be shared between different workers running in the same process. The same
+// interface could be implemented by something which shared these details with
+// a remote process via gRPC if that was desirable.
+// The goal is thatVersionedMapContents can be subsumed into this class - there
+// should be no reason to address it directly. For now, this class coordinates
+// writes, reads, and local state such that VersionedMapContents always contains
+// revisions which are no further ahead than the most recent SMR published by the
+// Map.
+type gossipHub struct {
+	contents *testonly.VersionedMapContents
+
+	mu            sync.RWMutex
+	pendingWrites map[uint64][]*trillian.MapLeaf
 
 	// SMRs are arranged from later to earlier (so [0] is the most recent), and the
 	// discovery of new SMRs will push older ones off the end. No SMR for a revision
 	// will be stored more than once.
-	smr [smrCount]*types.MapRootV1
+	smrs [smrCount]*types.MapRootV1
+}
+
+func newGossipHub() *gossipHub {
+	return &gossipHub{
+		contents:      &testonly.VersionedMapContents{},
+		pendingWrites: make(map[uint64][]*trillian.MapLeaf),
+	}
+}
+
+func (g *gossipHub) getLastReadRev() (rev uint64, found bool) {
+	if lastRoot := g.previousSMR(0); lastRoot != nil {
+		return lastRoot.Revision, true
+	}
+	return 0, false
+}
+
+// proposeLeaves should be called *before* writing a new map revision. The revision
+// will be available to readers only once advertiseSMR has been called with a root
+// that is >= the revision of this write.
+func (g *gossipHub) proposeLeaves(rev uint64, leaves []*trillian.MapLeaf) error {
```

Comment:
Systems like KT solve this by making writes idempotent, i.e. for a revision N they always propose the same leaves. Should we have this property in the hammer?

Reply:
That possibility might exist. This error already existed in the old code anyway and we haven't yet encountered it so I'm not too worried.

In the situation you describe the hammer would try again with a prng seeded with the same value, so would actually generate the same set of leaves (`retryOp` has `seed := w.prng.Int63()` before the retry logic). But then again it will likely fail on `WriteLeaves` because the revision we're trying to write already exists and the `map_rpc_server` will object, even if the update is logically the same. Modifying that is a feature request far beyond the scope of this PR though.